src/QsCompiler/QsTargeting/CallGraphWalker.fs

```diff
@@ -0,0 +1,239 @@
+        adjoint <- false
+        controlled <- false
+        this.ExpressionWalker method
+        inCall <- false
+        this.ExpressionWalker arg
+
+    override this.onOperationCall(method, arg) =
+        this.HandleCall method arg
+
+    override this.onFunctionCall(method, arg) =
+        this.HandleCall method arg
+
+    override this.onAdjointApplication(ex) =
+        adjoint <- true
+        base.onAdjointApplication(ex)
+
+    override this.onControlledApplication(ex) =
+        controlled <- true
+        base.onControlledApplication(ex)
+
+    override this.onIdentifier(sym, typeArgs) =
+        match sym with
+        | GlobalCallable(name) ->
+            if inCall
+            then
+                let kind = match adjoint, controlled with
+                           | false, false -> QsBody
+                           | false, true  -> QsControlled
+                           | true,  false -> QsAdjoint
+                           | true,  true  -> QsControlledAdjoint
+                graph.AddDependency(spec, name, kind, typeArgs)
+            else
+                // The callable is being used in a non-call context, such as being
+                /// assigned to a variable or passed as an argument to another callable,
+                // which means it could get a functor applied at some later time.
+                // We're conservative and add all 4 possible kinds.
+                graph.AddDependency(spec, name, QsBody, typeArgs)
+                graph.AddDependency(spec, name, QsControlled, typeArgs)
+                graph.AddDependency(spec, name, QsAdjoint, typeArgs)
+                graph.AddDependency(spec, name, QsControlledAdjoint, typeArgs)
+        | _ -> ()
+        base.onIdentifier(sym, typeArgs)
+
+and private ExpressionGraphBuilder(graph : CallGraph, spec : QsSpecialization) as this =
+    inherit ExpressionWalker()
+
+    let kindXformer = new ExpressionKindGraphBuilder(this, graph, spec)
+
+    override this.Kind = upcast kindXformer
+
+type private StatementGraphBuilder(scopeXformer : ScopeGraphBuilder, graph : CallGraph, 
+        spec : QsSpecialization) =
+    inherit StatementKindWalker()
+
+    let exprXformer = new ExpressionGraphBuilder(graph, spec)
+
+    override this.ScopeWalker x = scopeXformer.Walk x
+
+    override this.ExpressionWalker x = exprXformer.Walk x
+    override this.TypeWalker x = ()
+    override this.LocationWalker x = ()
+
+and private ScopeGraphBuilder(graph : CallGraph, spec : QsSpecialization) as this =
+    inherit ScopeWalker()
+
+    let kindXformer = new StatementGraphBuilder(this, graph, spec)
+
+    override this.StatementKind = upcast kindXformer
+
+    member this.Holder with get() = graph
+
+/// This syntax tree transformer fills in the CapabilityLevel fields in specializations,
+/// based on information gathered by the associated scope and other transformations.
+type TreeGraphBuilder() =
+    inherit SyntaxTreeWalker()
+
+    let graph = new CallGraph()
```

Comment:
I am not sure how you get the graph back to the calling context. I make a new TreeGraphBuilder obj, and use its Walk method to process my syntax tree, but where do I get back the graph?

Reply:
Argh... I messed that up while refactoring (the call graph walker was originally embedded into the capability leveler). I added the appropriate property, and also cleaned up some old/dead code.